misc/python/materialize/checks/all_checks/retain_history.py

```diff
@@ -0,0 +1,333 @@
+
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            schemas()
+            + dedent(
+                """
+                > CREATE TABLE time (time_index INT, t TIMESTAMP);
+                > INSERT INTO time VALUES (0, now());
+
+                # We want to have the topic created at a slightly later timestamp
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
+
+                $ kafka-create-topic topic=retain-history
+
+                $ kafka-ingest format=avro key-format=avro topic=retain-history key-schema=${keyschema} schema=${schema} repeat=5
+                {"key1": "K${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+
+                # Give the source some time
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
+
+                > INSERT INTO time VALUES (1, now());
+
+                > CREATE SOURCE retain_history_source
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-retain-history-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE UPSERT
+                  WITH (RETAIN HISTORY FOR '30s')
+
+                # Give the source some time
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
+
+                > INSERT INTO time VALUES (2, now());
+            """
+            )
+        )
+
+    def manipulate(self) -> list[Testdrive]:
+        return [
+            Testdrive(schemas() + dedent(s))
+            for s in [
+                """
+                $ kafka-ingest format=avro key-format=avro topic=retain-history key-schema=${keyschema} schema=${schema} repeat=2
+                {"key1": "K${kafka-ingest.iteration}"} {"f1": "B${kafka-ingest.iteration}"}
+
+                # Give the source some time
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
+
+                > INSERT INTO time VALUES (3, now());
+                """,
+                """
+                $ kafka-ingest format=avro key-format=avro topic=retain-history key-schema=${keyschema} schema=${schema} repeat=6
+                {"key1": "K${kafka-ingest.iteration}"} {"f1": "C${kafka-ingest.iteration}"}
+
+                # Give the source some time
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
+
+                > INSERT INTO time VALUES (4, now());
+
+                $ kafka-ingest format=avro key-format=avro topic=retain-history key-schema=${keyschema} schema=${schema} repeat=1
+                {"key1": "K${kafka-ingest.iteration}"} {"f1": "D${kafka-ingest.iteration}"}
+
+                # Give the source some time
+                $ sleep-is-probably-flaky-i-have-justified-my-need-with-a-comment duration="1500ms"
```

Comment:
Is there an alternative to this? Can I configure the source to update more frequently?

Reply:
What happens without the sleep?

Reply:
The results are not stable. I will sometimes get the data from before. Either the source updates only every one second or so or this is a product error... 🤔